src/python/pants/util/memo.py

```diff
@@ -24,21 +24,37 @@ def equal_args(*args, **kwargs):
   return key
 
 
+class InstanceKey(object):
+  """An equality wrapper for an arbirary object instance.
+
+  This wrapper leverages `id` and `is` for fast `__hash__` and `__eq__` but both of these rely on
+  the object in question not being gc'd since both `id` and `is` rely on the instance address which
+  can be recycled; so we retain a strong reference to the instance to ensure no recycling can occur.
+  """
+
+  def __init__(self, instance):
+    self._instance = instance
```

Comment:
This seems like a perfect spot for a `weakref`: https://docs.python.org/2/library/weakref.html , since we're already wrapping, and since the wrapper can continue to answer the relevant questions even after the weakref no longer resolves: ie, "is this equal" (if the weakref is dead, it's definitely not equal).

Reply:
...in fact, it looks like literally just using a `weakref` as the wrapper would have the right behaviour here:
> Weak references support tests for equality, but not ordering. If the referents are still alive, two references have the same equality relationship as their referents (regardless of the callback). If either referent has been deleted, the references are equal only if the reference objects are the same object.